compiler/qsc_passes/src/capabilitiesck.rs

```diff
@@ -0,0 +1,538 @@
+    UseOfDynamicDouble(#[label] Span),
+
+    #[error("cannot use a dynamic qubit")]
+    #[diagnostic(help(
+        "using a dynamic qubit, a qubit whose allocation depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicQubit"))]
+    UseOfDynamicQubit(#[label] Span),
+
+    #[error("cannot use a dynamic big integer value")]
+    #[diagnostic(help(
+        "using a dynamic big integer value, a big integer value that depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicBigInt"))]
+    UseOfDynamicBigInt(#[label] Span),
+
+    #[error("cannot use a dynamic string value")]
+    #[diagnostic(help(
+        "using a dynamic string value, a string value that depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicString"))]
+    UseOfDynamicString(#[label] Span),
+
+    #[error("cannot use a dynamically-sized array")]
+    #[diagnostic(help(
+        "using a dynamically-sized array, an array whose size depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicallySizedArray"))]
+    UseOfDynamicallySizedArray(#[label] Span),
+
+    #[error("cannot use a dynamic user-defined type")]
+    #[diagnostic(help(
+        "using a dynamic user-defined type, a user-defined type in which one or more of its members depend on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicUdt"))]
+    UseOfDynamicUdt(#[label] Span),
+
+    #[error("cannot use a dynamic function")]
+    #[diagnostic(help(
+        "using a dynamically resolved function, a function whose resolution depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicArrowFunction"))]
+    UseOfDynamicArrowFunction(#[label] Span),
+
+    #[error("cannot use a dynamic operation")]
+    #[diagnostic(help(
+        "using a dynamically resolved operation, an operation whose resolution depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicArrowOperation"))]
+    UseOfDynamicArrowOperation(#[label] Span),
+
+    #[error("cannot call a cyclic function with a dynamic value as argument")]
+    #[diagnostic(help(
+        "calling a cyclic function with a dynamic value as argument, an argument value that depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.CallToCyclicFunctionWithDynamicArg"))]
+    CallToCyclicFunctionWithDynamicArg(#[label] Span),
+
+    #[error("cannot define a cyclic operation specialization")]
+    #[diagnostic(help("cyclic operation specializations, specializations that contain cycles, are not supported by the target"))]
```

Comment:
maybe something like this would add more context for the user:

```suggestion
    #[diagnostic(help("operation specializations that contain call cycles are not supported by the current target"))]
```